core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala

```diff
@@ -111,94 +114,12 @@ class TiBatchWriteTable(
-        } else {
-          if (allNullOnAutoIncrement) {
-            df = df.drop(autoIncrementColName)
-            false
-          } else {
-            true
-          }
-        }
-      }
-
-      // when auto increment column is provided but the corresponding column in df contains null,
-      // we need throw exception
-      if (isProvidedID) {
-        autoIncProvidedID = true
-
-        if (!options.replace) {
-
-          val colNames =
-            tiTableInfo.getColumns.asScala.map(col => col.getName).mkString(", ")
-          throw new TiBatchWriteException(
-            s"""currently user provided auto increment value is only supported in update mode!
-               |please set parameter replace to true!
-               |
-               |colsInDf.length = ${colsInDf.length}
-               |
-               |df.schema = ${df.schema}
-               |
-               |tableColSize = $tableColSize
-               |
-               |colNames = $colNames
-               |
-               |tiTableInfo = $tiTableInfo
-            """.stripMargin)
-        }
-
-        if (!colsInDf.contains(autoIncrementColName)) {
-          throw new TiBatchWriteException(
-            "Column size is matched but cannot find auto increment column by name")
-        }
-
-        val hasNullValue = !df
-          .select(autoIncrementColName)
-          .rdd
-          .filter { row => row.get(0) == null }
-          .isEmpty()
-        if (hasNullValue) {
-          throw new TiBatchWriteException(
-            "cannot allocate id on the condition of having null value and valid value on auto increment column")
-        }
-        df.rdd
-      } else {
-        // if auto increment column is not provided, we need allocate id for it.
-        // adding an auto increment column to df
-        val newDf = df.withColumn(autoIncrementColName, lit(null).cast("long"))
-        val rowIDAllocator = getRowIDAllocator(count)
-
-        // update colsInDF since we just add one column in df
-        colsInDf = newDf.columns.toList.map(_.toLowerCase())
-        // last one is auto increment column
-        newDf.rdd.zipWithIndex.map { row =>
-          val rowSep = row._1.toSeq.zipWithIndex.map { data =>
-            val colOffset = data._2
-            if (colsMapInTiDB.contains(colsInDf(colOffset))) {
-              if (colsMapInTiDB(colsInDf(colOffset)).isAutoIncrement) {
-                val index = row._2 + 1
-                rowIDAllocator.getAutoIncId(index)
-              } else {
-                data._1
-              }
-            }
-          }
-          Row.fromSeq(rowSep)
-        }
-      }
+    val rdd = if (tiTableInfo.hasAutoIncrementColumn && isNeedAllocateIdForAutoIncrementCol(df)) {
+      // auto increment
+      allocateIdForAutoIncrementCol(count, startTimeStamp)
+    } else if (tiTableInfo.hasAutoRandomColumn && isNeedAllocateIdForAutoRandomCol(df)) {
+      // auto random
+      allocateIdForAutoRandomCol(count, startTimeStamp)
```

Comment:
Can we merged allocateIdForAutoIncrementCol and allocateIdForAutoRandomCol into one method with `type` parameter.

Reply:
I try to do so.PLAT